File: leetcode/reorder_list.py
# Given a singly linked list L: L0 -> .. -> Ln
# reorder it to: L0 -> Ln -> L1 -> Ln-1 ...
# Time Complexity: O(n)
# Space Complexity: O(1)
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # @param {ListNode} head
    # @return {void} Do not return anything, modify head in-place instead.
    def reorderList(self, head):
        if not head: return head
        fast = slow = head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
        slow.next, slow, prev = None, slow.next, None
        while slow:
            prev, slow.next, slow = slow, prev, slow.next
        while head and prev:
            prev.next,prev,head.next,head = head.next,prev.next,prev,head.next

if __name__ == "__main__":
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    solution = Solution()
    solution.reorderList(head)
    print head.val, head.next.val, head.next.next.val, head.next.next.next.val

File: leetcode/reorder_list.js
/* Given a singly linked list L0 -> L1 ...
 * Reorder to L0 -> Ln ...
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
 
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {void} Do not return anything, modify head in-place instead.
 */
var reorderList = function(head) {
    if (!head)
        return;
    for (var fast = head, slow = head; fast && fast.next; ) {
        slow = slow.next;
        fast = fast.next.next;
    }
    var temp = slow.next;
    slow.next = null;
    slow = temp;
    for (var prev = null; slow;) {
        temp = slow.next;
        slow.next = prev;
        prev = slow;
        slow = temp;
    }
    while (head && prev) {
        var temp1 = prev.next;
        var temp2 = head.next;
        prev.next = head.next;
        head.next = prev;
        prev = temp1;
        head = temp2;
    }
};
